TagsEdit code improvements

libklein opened the pull request:
Fix bug where pressing home on empty tag field crashes the program.

Bug is caused by https://github.com/libklein/keepassxc/blob/develop/src/gui/tag/TagsEdit.cpp#L340. The `currentText().isEmpty()` allows to invalidate the invariant if the current (and only) tag is empty. In this case, tags is resized to 0 (the only tag is erased) and the program crashes as `currentText()` accessed in https://github.com/libklein/keepassxc/blob/develop/src/gui/tag/TagsEdit.cpp#L504 fails as the tags list is now empty (https://github.com/libklein/keepassxc/blob/develop/src/gui/tag/TagsEdit.cpp#L387).

- ✅ Bug fix (non-breaking change that fixes an issue)

droidmonkey commented on src/gui/tag/TagsEdit.cpp:
```diff
@@ -337,7 +337,7 @@ struct TagsEdit::Impl
             std::count_if(tags.cbegin(), tags.cend(), [this](const auto& tag) { return tag.text == currentText(); });
-        if (currentText().isEmpty() || occurrencesOfCurrentText > 1) {
+        if ((currentText().isEmpty() && tags.size() > 1) || occurrencesOfCurrentText > 1) {
             tags.erase(std::next(tags.begin(), std::ptrdiff_t(editing_index)));
```
Yikes this line is ugly as sin. The line below this is what actually causes the crash because it decrements the editing index to -1. We also need to put guards in place in the `currentText` functions that blindly take the `editing_index` without any bounds checks.

libklein replied:
The change fixes the specific problem. The tags list will have at least size of 1, such that even after running the function, `editing_index` will be `>=0` and the invariant hold. 

Adding guards is probably a good idea, but I'd suggest to do some refactoring instead: What would be a good fallback for currentText() if editing_index is OoB? Crash gracefully? `currentRect` and `currentText` may return a reference.

droidmonkey replied:
currentText should return an empty string and currentRect a zero sized rect. At the end of the day the variables in this class are not safe and the class needs to be refactored to make it safer.

libklein replied:
Given the current design of the class, I'd consider handling out-of-bounds as follows:
If tags is not empty, set editing_index to 0, then return tags[editing_index].
If tags is empty, add an empty tag, set editing_index to 0 and return.

Alternatively, we could introduce set/get currentText()/currentRect() member functions to avoid the reference problem (`currentText()` and `currentRect()` return references that are actually written to https://github.com/libklein/keepassxc/blob/develop/src/gui/tag/TagsEdit.cpp#L399. This would be my preferred option.

droidmonkey replied:
That's just nuts, we should never be returning a reference to a QString. QString is optimized in the backend. If you are inside the class you don't/shouldn't generally use setter/getter functions. Those are meant for external users of the class.

libklein commented:
I've attempted to do some basic refactoring of the class. I've moved responsibility for keeping track of the current item, as well as maintaining the invariant into a dedicated "TagManager" class. This class avoids the index-tracking issues by using a QLinkedList rather than a QList, that we way never invalidate any iterators.

I've further attempted to have a clear split in responsibility between `Impl` and `TagsEdit`. `Impl` contains the tag handling logic, `TagsEdit` listens to events and calls `Impl` functions accordingly. I suggest merging `Impl` and `TagsEdit`, the only reason for having `Impl` in the first place was, I assume, to have a stable header API (PImpl). 

I've further fixed a few minor bugs in the implementation:
* There was a corner case where an empty tag could be rendered when TagsEdit went out of focus with an empty tag selected
* Scrollbars were not updated after removing a tag
* Tags were not re-rendered (and re-layouted!) when ";" is pressed on an empty tag.

There is still a bug with rendering - the height hint of the textbox is wrong on first render - and a bit of cleanup needs to be done. I'd however like to get feedback on the general approach before polishing. 

Another open question is the behavior of typing ";", which currently finishes editing the current tag. I think it's a bit unintuitive that this always **finishes** the current tag, regardless of cursor position. I'd expect the tag to be split if I type ";" in the middle of a tag. 

What do you think? Any idea why the initial height hint may be off?

droidmonkey commented:
Thank you for this work! I do agree that typing a ; or , should cause the tags to be split instead of ending typing immediately. But what about pressing space, that also ends tag typing. That could also help with pasting tag strings.

libklein committed:
TagsEdit code improvements and crash fix

Fix crash when pressing home on empty tag field

Move completer to TagsEdit.
Move cursor blink status to TagsEdit.
Move paint implementation to impl.
Simplify calcRect and drawTag.

Hide editing_index and cursor position.
Fix bug where an empty tag was shown if the tag edit was unfocused.
Fix a bug where the scollbar was not updated when a tag was removed.

Hide remaining TextEdit internal fields.
Refactor to use QLinkedList.
Remove obsolete EmptyTagIterator.
Encapsulate tags and selected index in tags manager.

droidmonkey commented:
@libklein sorry just coming back to this. I made a bunch of code improvements to align with our standards. Also noticed that now the tags view when editing an entry gets cut off when first shown:

![image](https://github.com/user-attachments/assets/24758a17-d57a-4571-a3bd-d0c7cf3f8476)

The view expands vertically when focused to not be cut off. Not sure where in the list of changes this happened.

michaelk83 commented:
1. Looks like too much vertical padding on the tag labels (inside the green rectangles).
2. Could reduce at least the vertical padding of the outer text box (outside the green rectangles).
3. The text box is shifted inward from its gray border, which makes the text box less tall, and shifts it sideways. (On closer look, the other textboxes have that border too, but the black area is not shifted.)

droidmonkey commented:
@michaelk83 those are not the causes. I put a bunch of time into diagnosing this and we were basically fighting Qt for sizing. I have a WIP commit added that improves the situation substantially and allows for external styling of the tags display instead of overriding everything internally.

![image](https://github.com/user-attachments/assets/eab20c0a-8934-471a-9046-1caf29aad584)

![image](https://github.com/user-attachments/assets/1f838405-8dea-476b-827c-585d4dc9de0b)

~~Note: This still needs work because editing and adding tags is now broken. The fix is to separate actual painting of the widget from height estimation, the two are stuck together now and the logic is broken.~~ FIXED

droidmonkey committed:
Enable proper sizing and layout of the tags display

droidmonkey commented:
TODO:

- [ ] Combine the TagsEdit::Impl directly into TagsEdit. There is no reason to have these separate and it creates ugly code
- [ ] TagsManager should do all things non-gui related to the tags, including non-gui tag editing functions
- [ ] Make a proper viewport height calculation function

droidmonkey committed:
WIP: Cleanup TagsEdit.cpp

droidmonkey commented:
Going to save this refactor for 2.8.0, created a new compact crash fix for 2.7.10: https://github.com/keepassxreboot/keepassxc/pull/11625

libklein commented:
@droidmonkey Sorry, I did not have time to look into this since December. 

If we have time to do a large refactor, then I'd suggest replacing the apparently brittle custom rendering logic altogether. Is there any reason to not implement TagsEdit as a container that nests custom QTextEdits? That would allow to have rich text support etc.

droidmonkey commented:
I was thinking the same thing @libklein. There is plenty of time on this branch/PR since it is targeting 2.8.0 now. We merged the break/fix for the HOME button crash already.
